[MERGE][ADD] (sale,purchase,account)(_edi*): add EDI for orders

kcv-odoo opened the pull request:
This PR introduces the ability to create Sales Orders from Purchase Orders, through the standard UBL BIS 3 format.
It allows an Odoo user to easily import his purchase orders from customers (using Odoo too) into his own instance, with the SO being automatically filled with the matching data (if correctly configured).  If the information cannot be properly matched with his own data, they'll be requested to fill the missing information.

Enterprise PR: odoo/enterprise#66469

task-3791197

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/172450.png)](https://mergebot.odoo.com/odoo/odoo/pull/172450)

Feyensv commented:
Ready for review so that accounting & logicstics team can take a look (most of the changes impact their code and logic)

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1709,6 +1795,104 @@ def _get_default_payment_link_values(self):
+        """
+        self.ensure_one()
+
+        def close_file(file_data):
+            if file_data.get('on_close'):
+                file_data['on_close']()
```
why do you need a method for that? :thinking:

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -612,6 +628,56 @@ def _compute_amount_undiscounted(self):
+            'state',
+            'amount_total',
+        )
+        self.env["sale.order"].flush_model(used_fields)
```
```suggestion
        self.env['sale.order'].flush_model(used_fields)
```

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+        return True
+
+    def _import_order_lines(self, order, tree, xpath):
+        """ Imoprt order lines from xml tree.
```
```suggestion
        """ Import order lines from xml tree.
```

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+            # To do: rename product_uom field to `product_uom_id` of sale.order.line
+            del line_values['product_uom_id']
+            if not line_values['product_id']:
+                logs += [_("Could not retrive product for line '%s'", line_values['name'])]
```
```suggestion
                logs += [_("Could not retrieve product for line '%s'", line_values['name'])]
```

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+            del line_values['product_uom_id']
+            if not line_values['product_id']:
+                logs += [_("Could not retrive product for line '%s'", line_values['name'])]
+            # To do: rename tax_id field to `tax_ids` of sale.order.line
```
not now, will be for post oxp

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+    def _import_delivery_partner(self, order, name, phone, email):
+        """ Import delivery address from details if not found then log details."""
+        logs = []
+        dest_partner = self.env['res.partner'] \
+            .with_company(order.company_id) \
+            ._retrieve_partner(name=name, phone=phone, email=email)
```
```suggestion
        dest_partner = self.env['res.partner'].with_company(
            order.company_id
        )._retrieve_partner(name=name, phone=phone, email=email)
```

Please avoid line breaks with `\` :pray:

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+    def _get_partner_detail_str(self, name=False, phone=False, email=False, vat=False):
+        """ Return partner details string to help user find or create proper contact with details.
+        """
+        partner_details = ""
+        if name:
+            partner_details += "\n Name: %s" % name
+        if phone:
+            partner_details += ",\n Phone: %s" % phone
+        if email:
+            partner_details += ",\n Email: %s" % email
+        if vat:
+            partner_details += ",\n Vat: %s" % vat
```
translation for non english speakers ?

kcv-odoo replied:
Already added translation for whole string where it used :see_no_evil: 
Do we need to add translation in each sub string to make it properly work I think I need to add kw string instead of %s to not have weird string :thinking:

Feyensv replied:
```python
a = 'hello world'
b = 'weather's nice today'

c = _("%s %s", a, b) # Doesn't work

a = _("hello world")
b = _("weather's nice today")
c = _("%s %s", a, b) # works, but doesn't allow translators to fully reorganize the sentence according to language specificities
c = _("%(a)s %(b)s", a=a, b=b)  # more control, translators can move b before a if needed, but cannot split the content of a or b.

c = _("hello world weather's nice today")  # Gives translators full control on the whole sentence
```

More details and explanation [here](https://www.odoo.com/documentation/master/contributing/development/coding_guidelines.html#use-translation-method-correctly)

kcv-odoo replied:
Got it! Thank you for detailed explanation :smile:

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,126 @@
+
+        return dest_partner, logs
+
+    def _import_partner(self, company_id, name, phone, email, vat,
+        country_code=False,
+        peppol_eas=False,
+        peppol_endpoint=False,
+    ):
```
```suggestion
    def _import_partner(self, company_id, name, phone, email, vat, **kwargs):
```

Feyensv commented on addons/sale_edi_ubl/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+
+        :return: None.
+        """
+        activity_message = _(
+            "Some information could not be imported"
+        )
```
```suggestion
        activity_message = _("Some information could not be imported")
```

Feyensv commented on addons/sale_edi_ubl/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+    def _get_line_vals_list(self, lines_vals):
+        """ Get sale order line values list.
+
+        param list line_vals: List of values [name, qty, price, tax].
```
```suggestion
        :param list line_vals: List of values [name, qty, price, tax].
```

Feyensv commented on addons/sale_edi_ubl/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    # -------------------------------------------------------------------------
+    # EDI
+    # -------------------------------------------------------------------------
```
```suggestion
```

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1709,6 +1795,104 @@ def _get_default_payment_link_values(self):
+                    file_name=file_data['filename'],
+                    decoder=decoder.__name__,
+                )
+                self.sudo().message_post(body=message)
```
why the sudo here?

kcv-odoo replied:
Because message should be logged from odoobot :thinking:

Feyensv replied:
That's not `sudo` for that, that's `with_user(SUPERUSER_ID)` :eyes:

kcv-odoo replied:
Yes I know :sweat_smile: 
But I also think may be in future we might create SO from email like invoice or actually use peppol end points on customer where sending doc of Purchase order file fetched and create Sale Order by system (I don't how actually peppol end points works :see_no_evil: ) :thinking: 
But I think I should use `with_user(SUPERUSER_ID)` for now :see_no_evil:

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1709,6 +1795,104 @@ def _get_default_payment_link_values(self):
+        if not attachments:
+            raise UserError(_("No attachment was provided"))
+
+        orders = self
```
```suggestion
        orders = self.browse()
```

Feyensv reviewed (commented):
Quick review for the `sale` module :duck:

JulienVR commented on addons/account_edi_ubl_cii/models/account_edi_common.py:
```diff
@@ -680,68 +682,71 @@ def _retrieve_fixed_tax(self, invoice_line, fixed_tax_vals):
-            if hasattr(invoice_line, '_predict_specific_tax'):
-                # company check is already done in the prediction query
-                predicted_tax_id = invoice_line \
-                    ._predict_specific_tax('percent', amount, invoice_line.move_id.journal_id.type)
-                tax = self.env['account.tax'].browse(predicted_tax_id)
+            if hasattr(record, '_get_specific_tax'):
+                tax = record._get_specific_tax(line_values['name'], 'percent', amount, tax_type)
```
can't we simply do
`if hasattr(record, '_predict_specific_tax'):`
and keep the existing lines as is ? we could then remove `_get_specific_tax`

JulienVR replied:
Also see my other comment in the enterprise part: I'm not sure `_predict_specific_tax` should be api.model

kcv-odoo replied:
To be able to avoid creating lines to predict specific tax we are making `_predict_specific_tax` model method

JulienVR reviewed (commented):
Hello !

Thanks for your work on this one :)

I don't really know purchase / sale. Maybe we could meet this week to discuss the new features ? :) There are lots of new JS files...

For `account_edi_ubl_cii` I don't see any red flag now, and the tests in `l10n_account_edi_ubl_cii_tests` are green so it's a good beginning. :+1:

kcv-odoo commented on addons/account_edi_ubl_cii/data/ubl_20_templates.xml:
```diff
@@ -2,8 +2,7 @@
     <template id="ubl_20_AddressType">
-        <t xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
-           xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
+        <t xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
```
Yes, To be able to refactor purchase export and use this templates there in future :see_no_evil:

lordkrandel commented on addons/purchase_edi_ubl_bis3/__manifest__.py:
```diff
@@ -1,6 +1,5 @@
 {
     'name': "Import/Export electronic orders with UBL",
-    'version': '1.0',
```
Reason for removing the version of the module?

Feyensv replied:
Probably because I asked for the removal in the sale module ^^.
Specifying 1.0 is useless since it's the default manifest value, but it doesn't have to be removed from the existing code indeed.

kcv-odoo replied:
Yes my bad I removed it from here instead sale edi :cry:

lordkrandel commented on addons/sale/models/sale_order.py:
```diff
@@ -1695,6 +1781,101 @@ def _get_default_payment_link_values(self):
+        :rtype: dict
+        """
+        orders = self._create_order_from_attachment(attachment_ids)
+        action_vals = {
+            'name': _("Generated Orders"),
+            'domain': [('id', 'in', orders.ids)],
+            'res_model': 'sale.order',
+            'type': 'ir.actions.act_window',
+            'context': self._context
+        }
+        if len(orders) == 1:
+            action_vals.update({
+                'views': [[False, 'form']],
+                'res_id': orders.id,
+            })
+        else:
+            action_vals.update({
+                'views': [[False, 'list'], [False, 'kanban'], [False, 'form']],
+            })
+
+        return action_vals
```
Check out `odoo/odoo/addons/base/models/ir_ui_view.py`, there is some `Model._get_records_action` 

![immagine](https://github.com/user-attachments/assets/bf8d0031-77a2-4abe-b88b-509c1d96708e)

kcv-odoo replied:
Got it!

kcv-odoo commented:
> I don't really know purchase / sale. Maybe we could meet this week to discuss the new features ? :) There are lots of new JS files...

Actually I am from India so may be we can do meeting online also I had initial discussion with @smetl about re-using accounting EDI :see_no_evil:  and in JS code I just extracted upload functionality from Invoice and split files (https://www.odoo.com/odoo/project/967/tasks/3985140) to be able re-use that for Sale Order to avoid code duplication

rugo-odoo commented on addons/account/static/src/components/document_file_uploader/document_file_uploader.js:
```diff
@@ -0,0 +1,77 @@
+            }
+            delete action.context.notifications;
+        }
+        if (action.help && typeof(action.help) === "string") {
```
```suggestion
        if (action.help?.length) {
```
I think this will do the same.

Also, if I'm not mistaken, we just want non empty strings, this condition is not very clear about that imo.

rugo-odoo commented on addons/account/static/src/components/document_file_uploader/document_file_uploader.js:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    get divClass() {
+        return this.props.divClass || (this.props.record && this.props.record.data ? 'oe_kanban_color_' + this.props.record.data.color : '');
```
```suggestion
        return this.props.divClass || (this.props.record && this.props.record.data ? `oe_kanban_color_${this.props.record.data.color}` : '');
```

rugo-odoo commented on addons/account/static/src/components/bill_guide/bill_guide.scss:
```diff
@@ -0,0 +1,25 @@
+.account_drag_drop_btn {
+    border-style: dashed !important;
+    border-color: $o-brand-primary;
+    background-color: #F2EDF0;
```
We should use an existing variable instead.

kcv-odoo replied:
I don't know there is existing code for this :thinking:

rugo-odoo commented on addons/account/static/src/views/account_dashboard_kanban/account_dashboard_kanban.scss:
```diff
@@ -0,0 +1,14 @@
+        border: 1px dashed $border-color;
+        &.drag_to_card {
+            border-color: $o-brand-primary;
+            background-color: #F2EDF0;
```
We should using an existing variable.

rugo-odoo commented on addons/account/static/src/views/account_dashboard_kanban/account_dashboard_kanban_renderer.js:
```diff
@@ -0,0 +1,44 @@
+        return false;
+    }
+
+    disableDragging() {
+        this.env.dashboardState.isDragging = false;
+    }
```
```suggestion
    setDragging(value) {
        this.env.dashboardState.isDragging = value;
    }
```
What about something like this ?

We create the function to use 2 times for the disable but we also have 2 times for enable so why not do something like this instead ?

rugo-odoo commented on addons/account/static/src/views/upload_file_from_data_hook.js:
```diff
@@ -0,0 +1,68 @@
+                console.warn("Invalid files to extract details.");
+                return;
+            }
+            const selector = '.document_file_uploader.o_input_file';
+            let uploadInput = document.querySelector(selector);
```
```suggestion
            let uploadInput = document.querySelector('.document_file_uploader.o_input_file');
```
Don't think we need the variable.

kcv-odoo commented:
Hello @rugo-odoo :wave:

Applied most of the changes changes!
About your questions I don't fully explore css I just move the css code dedicated file so may you should ask @h4818 if you have doubts regarding choices of css :thinking: 

Have a nice day!

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -611,6 +627,56 @@ def _compute_amount_undiscounted(self):
+        :rtype: dict
+        """
+        orders = self.filtered('client_order_ref')
+        if not orders:
```
```suggestion
        if not orders.filtered('id'):
```
Even if a ref is given, we shouldn't search duplicated orders for orders which are not saved.

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1004,9 +1066,23 @@ def _should_be_locked(self):
+                    "The following orders are not in a state requiring confirmation: %s",
+                    ", ".join(self.mapped('display_name')),
+                ))
+            if any(
+                not line.display_type
+                and not line.is_downpayment
+                and not line.product_id
+                for line in order.order_line
+            ):
+                raise ValidationError(_(
+                    "A line on these orders missing a product, you cannot confirm it."
+                ))
```
Isn't this already enforced by the SQL constraint on `sale.order.line` ?

kcv-odoo replied:
No it isn't, There is not constraint like that on SOL :thinking: 
Adding one would prevent extraction :see_no_evil:

Feyensv replied:
```python
        ('accountable_required_fields',
            "CHECK(display_type IS NOT NULL OR is_downpayment OR (product_id IS NOT NULL AND product_uom IS NOT NULL))",
            "Missing required fields on accountable sale order line."),
```
:thinking: 

Also, your changes here are probably the cause of the failing query count in `sale_subcription` tests, you have to update the query count (or find a way to avoid those additional queries when they're not necessary).

kcv-odoo replied:
I am not sure why that constraint is not triggering :slightly_frowning_face: 

Updating query count in `sale_subscription` (We really don't want to add this error here but it was allowing SO to confirm even some SOL does not have product field set on them also UX require attribute was also not triggering because sol is not in edition mode :cry: )

Feyensv replied:
Yeah, it's strange it's not working :confused:
Let's update of the query count for now.
Could you update the method so that it returns a boolean `_can_be_confirmed()`, and only raise the ValidationError in `action_confirm` if the returned value is falsy?  That way, other modules can override the check more easily ;).

kcv-odoo replied:
I did this to have different error message depending on issue
Is it okay to have same error message :thinking: 
(

Feyensv replied:
Could work :ok_hand: (if the strings are correctly translated :eyes:).

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1690,6 +1776,84 @@ def _get_default_payment_link_values(self):
+            })
+            order._extend_with_attachments(attachment)
+            orders |= order
+            order.message_post(attachment_ids=attachment.ids)
```
Why do you post a message here? 
Is it only to link the attachments to the chatter ? :thinking:
It seems strange to call `message_post` without any message

kcv-odoo replied:
Because adding file in chatter we are extracting data from
I guess we can add some message like `Extracted order from file` or something like that :thinking: 
Same thing is done in lot of other places only post attachments without message (e.g. in invoice)

Feyensv replied:
AFAIK, simply linking the file to the record (with res_id & res_model) is sufficient and there shouldn't be a need to call `message_post`.  We can live with that for now, but I believe it could be avoided in the future.

Feyensv commented on addons/sale_edi_ubl/models/sale_edi_common.py:
```diff
@@ -0,0 +1,124 @@
+            return False
+
+        return self.env['account.payment.term'].search([
+            *self.env['account.journal']._check_company_domain(order.company_id),
```
should be the same model as the model you're searching on :thinking:

Feyensv reviewed (commented):
Last minute comments and questions.

Feyensv commented:
Hello @JulienVR can you take another look to see if Kartik answers and changes are good for you?
As this is a priority for the freeze, I'd like to r+ it tomorrow or Wednesday, unless there are blocking points.

kcv-odoo commented on addons/account/controllers/download_docs.py:
```diff
@@ -60,3 +61,15 @@ def download_invoice_documents_filetype(self, invoices, filetype, allow_fallback
+                'content': response.content,
+                'content_type': response.headers.get('Content-Type'),
+                'file_name': file_name or "document",
+            }
```
Done!
It seemed important at first because on portal of purchase order were saying customer to copy/paste url to create so
![image](https://github.com/user-attachments/assets/09443b3e-7cea-4b23-aedf-aa0b6173c2d0)
But I get you point we need to check security concerns and also we can add this in stable 18.0 after freeze

Keeping copy/pasting files as it look useful so user can copy/paste files from file manager to here and it does not have any issue related to security and file types and also user can select bunch of files from there and paste it in list or kanban view
Still we can drop the remaining part from copy/paste commit as currently it isn't easy to focus on list/kanban view full of records as there is not space to focus except footer.

what do you think?

Feyensv replied:
Yeah, ofc, copy/paste and drag & drop are useful and must be kept.
As long as we make sure that the user has the rights to create the targeted record (and we make sure we do not bypass the existing rights)

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -611,6 +627,56 @@ def _compute_amount_undiscounted(self):
+                 AND sale_order.id != duplicate_order.id
+                 AND duplicate_order.state != 'cancel'
+                 AND sale_order.partner_id = duplicate_order.partner_id
+                 AND sale_order.amount_total = duplicate_order.amount_total
```
I don't think we should enforce the amounts to be equal.
What if I import a purchase order, then the client adds other products to the PO, I import the new version of the XML file, and I've two orders with the same customer reference, but different amounts (and I should ideally delete or cancel the first one).

kcv-odoo replied:
Done remove amount dependency
Also update purchase export little bit to have `client_order_ref` unique per customer depending on PO name instead partner_ref :see_no_evil:

Feyensv commented on addons/account/static/src/components/document_file_uploader/document_file_uploader.js:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    async onFileUploaded(file) {
+        const att_data = {
+            name: file.name,
+            mimetype: file.type,
+            datas: file.data,
+        };
+        const [att_id] = await this.orm.create("ir.attachment", [att_data], {
+            context: { ...this.extraContext, ...this.env.searchModel.context },
+        });
+        this.attachmentIdsToProcess.push(att_id);
```
All internal users are allowed to create attachments, but this doesn't mean that they are allowed to create invoice/SO/..., even if they have read rights on the model.
I believe this should be improved in the future (let's keep it like that for now, it was already like that)

Feyensv commented on addons/sale/models/sale_order.py:
```diff
@@ -1192,6 +1268,16 @@ def _get_action_add_from_catalog_extra_context(self):
     def _get_product_catalog_domain(self):
         return expression.AND([super()._get_product_catalog_domain(), [('sale_ok', '=', True)]])
 
+    def action_open_business_doc(self):
+        self.ensure_one()
+        return {
+            'name': _("Order"),
+            'type': 'ir.actions.act_window',
+            'res_model': 'sale.order',
+            'res_id': self.id,
+            'views': [(False, 'form')],
+        }
```
I don't see where this method is used :thinking:

kcv-odoo replied:
This is used in `x2many_buttons` widget to open form view of duplicated record
I think accounting team added this method in py for journal because they want to open different model form [depending on record data](https://github.com/odoo/odoo/blob/master/addons/account/models/account_move.py#L4670-L4693)

Feyensv commented:
@robodoo rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

AntoineVDV commented:
@robodoo override=ci/security
Reason: flagged because the code was moved in another file; see original explanation [here](https://github.com/odoo/odoo/pull/168037#issuecomment-2161335462)
Link to ci/security build (master): https://runbot.odoo.com/runbot/build/68380877

Feyensv commented:
@robodoo r+

robodoo commented:
@kcv-odoo @Feyensv unable to stage: merge conflict

AntoineVDV commented:
@robodoo r+
